File: bridge/src/main/java/ru/hollowhorizon/hollowengine/api/AutoModelType.java
package ru.hollowhorizon.hollowengine.api;

@FunctionalInterface
public interface AutoModelType {
    AutoModelType DEFAULT = () -> "item/generated";
    AutoModelType HANDHELD = () -> "item/handheld";
    AutoModelType CUBE_ALL = () -> "block/cube_all";

    static AutoModelType custom(String type, String blockState) {
        return new AutoModelType() {
            @Override
            public String modelId() {
                return type;
            }

            @Override
            public String blockStateId() {
                return blockState;
            }
        };
    }

    static AutoModelType custom(String type) {
        return custom(type, "default");
    }

    String modelId();

    default String blockStateId() {
        return "default";
    }
}

File: src/main/java/ru/hollowhorizon/hollowengine/common/registry/AutoModelType.kt
package ru.hollowhorizon.hollowengine.common.registry

/**
 * Possibility to select automatic model generation.
 */
fun interface AutoModelType {
    fun modelId(): String
    fun blockStateId(): String = "default"

    companion object {
        val DEFAULT = AutoModelType { "item/generated" }
        val HANDHELD = AutoModelType { "item/handheld" }
        val CUBE_ALL = AutoModelType { "block/cube_all" }
        fun custom(type: String, blockState: String = "default") = object : AutoModelType {
            override fun modelId() = type
            override fun blockStateId() = blockState
        }
    }
}
